content/data-structures/Treap.h

```diff
@@ -1,67 +1,85 @@
 /**
- * Author: someone on Codeforces
+ * Author: someone on Codeforces and then modified by Iván Renison
  * Date: 2017-03-14
  * Source: folklore
  * Description: A short self-balancing tree. It acts as a
- *  sequential container with log-time splits/joins, and
- *  is easy to augment with additional data.
+ *  sequential container with log-time splits, joins, queries ans updates.
+ *  Can also support reversals with the commented REVERSE lines
  * Time: $O(\log N)$
- * Status: stress-tested
+ * Status: stress-tested and problem tested
  */
 #pragma once
 
+typedef ll T; typedef ll L; // T: data type, L: lazy type
+const T tneut = 0; const L lneut = 0; // neutrals
+T f(T a, T b) { return a + b; } // operation
+// new st according to lazy
+T apply(T v, L l, ll len) { return v + l * len; }
+// cumulative effect of lazy
+L comb(L a, L b) { return a + b; }
+
 struct Node {
 	Node *l = 0, *r = 0;
-	ll val, y, c = 1;
-	Node(ll val) : val(val), y(rand()) {}
-	void recalc();
-};
-
-ll cnt(Node* n) { return n ? n->c : 0; }
-void Node::recalc() { c = cnt(l) + cnt(r) + 1; }
-
-template<class F> void each(Node* n, F f) {
-	if (n) { each(n->l, f); f(n->val); each(n->r, f); }
-}
-
-pair<Node*, Node*> split(Node* n, ll k) {
-	if (!n) return {};
-	if (cnt(n->l) >= k) { // "n->val >= k" for lower_bound(k)
-		auto pa = split(n->l, k);
-		n->l = pa.snd;
-		n->recalc();
-		return {pa.fst, n};
-	} else {
-		auto pa = split(n->r, k - cnt(n->l) - 1); // and just "k"
-		n->r = pa.fst;
-		n->recalc();
-		return {n, pa.snd};
+	T val, acc; L lazy = lneut;
+	ll y, c = 1;
+	// bool rev = false; // REVERSE
+	Node(T val = tneut) : val(val), acc(val), y(rand()) {}
+	void recalc() {
+		c = 1, acc = tneut;
+		if (l) l->push(), acc = f(acc, l->acc), c += l->c;
+		acc = f(acc, val);
+		if (r) r->push(), acc = f(acc, r->acc), c += r->c;
 	}
-}
-
-Node* merge(Node* l, Node* r) {
-	if (!l) return r;
-	if (!r) return l;
-	if (l->y > r->y) {
-		l->r = merge(l->r, r);
-		l->recalc();
-		return l;
-	} else {
-		r->l = merge(l, r->l);
-		r->recalc();
-		return r;
+	void push() {
+		// if (rev) { // REVERSE
+		//   swap(l, r), rev = false;
+		//   if (l) l->rev ^= 1; if (r) r->rev ^= 1;
+		// }
+		val = apply(val, lazy, 1), acc = apply(acc, lazy, c);
+		if (l) l->lazy = comb(l->lazy, lazy);
+		if (r) r->lazy = comb(r->lazy, lazy);
+		lazy = lneut;
+	}
+	ll cnt(Node* n) { return n ? n->c : 0; }
```

Comment:
```suggestion
	static ll cnt(Node* n) { return n ? n->c : 0; }
```

De esa forma se puede hacer Node::cnt(n) desde fuera de Node.

Reply:
Ahora que lo pienso, `cnt` solo se usa de una sola forma, así que lo podría sacar directamente